Hearthstone Deck Tracker/FlyoutControls/Options/OptionsSearch.xaml.cs

```diff
@@ -50,44 +51,111 @@ private List<CheckBoxWrapper> LoadWrappers()
-				checkBoxWrappers.AddRange(checkBoxes.Select(cb => new CheckBoxWrapper {MenuItem = option, CheckBox = cb}));
+				var option = Helper.FindLogicalDescendants<DependencyObject>(IsSearchableOption, optionsMenuItem.UserControl);
+				optionWrappers.AddRange(option.Select(cb => WrapOption(optionsMenuItem, cb)));
 			}
-			return checkBoxWrappers;
+			return optionWrappers;
+		}
+
+		private bool IsSearchableOption(DependencyObject depObj)
+		{
+			var basicTypes = new List<Type>{typeof(CheckBox)};
+			if(basicTypes.Contains(depObj.GetType())) return true;
+			if(depObj is DockPanel)
+			{
+				var dockPanel = depObj as DockPanel;
+				foreach(var child in dockPanel.Children)
+				{
+					if(basicTypes.Contains(child.GetType())) return true;
+				}
+			}
+			return false;
+		}
+
+		private OptionWrapper WrapOption(UserControlWrapper menuItem, DependencyObject depObj)
+		{
+			if(depObj is DockPanel)
+			{
+				var panel = depObj as DockPanel;
+				foreach(var child in panel.Children)
+				{
+					if(child is CheckBox)
+						return WrapSimpleOption(menuItem, child as CheckBox);
+				}
+			}
+			return WrapSimpleOption(menuItem, depObj as ContentControl);
+		}
+
+		private OptionWrapper WrapSimpleOption(UserControlWrapper menuItem, ContentControl control)
+		{
+			if(control == null)
+				throw new ArgumentNullException("control");
+			if(control is CheckBox)
+				return new CheckBoxWrapper(menuItem, control as CheckBox);
+			throw new ArgumentException("Argument must be a wrappable option type.", "control");
 		}
 
 		private void UpdateSearchResult(string text)
 		{
 			ListBoxSearchResult.Items.Clear();
 			if(string.IsNullOrEmpty(text))
 				return;
-			foreach(var wrapper in CheckBoxWrappers.Where(x => x.CheckBox.Content.ToString().ToUpperInvariant().Contains(text.ToUpperInvariant())
-															|| (x.CheckBox.Name.ToUpperInvariant().Replace("CHECKBOX", "").Contains(text.ToUpperInvariant()))))
+			foreach(var wrapper in OptionWrappers.Where(x => x.Matches(text)))
 				ListBoxSearchResult.Items.Add(wrapper);
 		}
 
 		private void ListBoxSearchResult_OnMouseDoubleClick(object sender, MouseButtonEventArgs e)
 		{
-			var selected = (sender as ListBox)?.SelectedItem as CheckBoxWrapper;
+			var selected = (sender as ListBox)?.SelectedItem as OptionWrapper;
 			if(selected == null)
 				return;
-			var tvis = Helper.FindLogicalChildren<TreeViewItem>(Core.MainWindow.Options.TreeViewOptions);
+			var tvis = Helper.FindLogicalDescendantsDeep<TreeViewItem>(Core.MainWindow.Options.TreeViewOptions);
 			var target = tvis.FirstOrDefault(x => x.Name.Contains(selected.MenuItem.Name.Substring(7)));
 			if(target != null)
 			{
-				if(selected.CheckBox.Visibility == Visibility.Collapsed)
+				if(selected.OptionControl.Visibility == Visibility.Collapsed)
 					AdvancedOptions.Instance.Show = true;
 				target.IsSelected = true;
 			}
 		}
 
-		public class CheckBoxWrapper
+		public abstract class OptionWrapper
```

Comment:
I'd prefer to make this a generic `public abstract class OptionWrapper<T>` because we really want to parameterize `OptionControl`. For instance, `CheckBoxWrapper` would be an `OptionWrapper<CheckBox>` whose `OptionControl` is already a `CheckBox`, so it wouldn't need a separate `CheckBox` property that refers to the same control as `OptionControl`.

But I couldn't make that work with collections. eg, I couldn't find a way to do this:

```
private List<OptionWrapper<T>> LoadWrappers()
{
  var optionWrappers = new List<OptionWrapper<T>>; 
  // ...  
}
```

Any thoughts? I haven't touched C# in 9 years...